flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala

```diff
@@ -176,7 +184,15 @@ object GenerateUtils {
     val resultTerm = ctx.addReusableLocalVariable(resultTypeTerm, "result")
     val isResultNullable = resultNullable || (isReference(returnType) && !isTemporal(returnType))
     val nullTermCode = if (ctx.nullCheck && isResultNullable) {
-      s"$nullTerm = ($resultTerm == null);"
+      // we assume that result term of a primitive SQL type will never be null,
+      // violating this assumption might cause null pointer exception somewhere else.
+      // when using this column we'll first check its null term.
+      s"""
+         |$nullTerm = ($resultTerm == null);
```

Comment:
I'm worried about adding a few lines of code here. This is the only way for each call, which will increase the code generated by CodeGen.

Can we just fix the result term or add null check in the `CalcCodeGenerator`?

Reply:
Indeed. I used to think that this change will only affect non-primitive numbers and booleans. However string is also a non-primitive type so this change might affect a larger scope than I thought.

I once also thought of checking null terms when generating filter code, however this will affect almost all generated code which I also think is not efficient.

Currently I can't come up with other places where null terms aren't checked before result terms are used. So I'd like to only change the `GeneratedExpression` produced by casting to boolean.